cvat/apps/engine/cloud_provider.py

```diff
@@ -196,29 +206,30 @@ def optimally_image_download(self, key: str, chunk_size: int = 65536) -> BytesIO
     def bulk_download_to_memory(
         self,
         files: List[str],
-        threads_number: int = min(get_cpu_number(), 4),
+        *,
+        threads_number: Optional[int] = None,
         _use_optimal_downloading: bool = True,
-    ) -> List[BytesIO]:
+    ) -> Iterator[BytesIO]:
         func = self.optimally_image_download if _use_optimal_downloading else self.download_fileobj
-        if threads_number > 1:
-            with ThreadPool(threads_number) as pool:
-                return pool.map(func, files)
-        else:
-            slogger.glob.warning('Download files to memory in series in one thread.')
-            return [func(f) for f in files]
+        threads_number = normalize_threads_number(threads_number, len(files))
+
+        with ThreadPoolExecutor(max_workers=threads_number) as executor:
+            yield from executor.map(func, files)
 
     def bulk_download_to_dir(
         self,
         files: List[str],
         upload_dir: str,
-        threads_number: int = min(get_cpu_number(), 4),
-    ):
+        *,
+        threads_number: Optional[int] = None,
+    ) -> None:
+        threads_number = normalize_threads_number(threads_number, len(files))
+
         args = zip(files, [os.path.join(upload_dir, f) for f in files])
         if threads_number > 1:
             with ThreadPool(threads_number) as pool:
-                return pool.map(lambda x: self.download_file(*x), args)
+                pool.map(lambda x: self.download_file(*x), args)
```

Comment:
The `map()` function returns a generator. The results can be missing when the code here exits. Consider adding a `list()` call for the map result.

https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Executor.map

Reply:
Honestly, it is not true since `ThreadPool` was used here, but okay, I switched to `ThreadPoolExecutor`

Reply:
Ok, you're right, I checked the wrong type. The updated version looks ok to me, but I'd probably just use a `map` call, as it was earlier, for simplicity.

Reply:
> but I'd probably just use a map call, as it was earlier, for simplicity

I've used `map` in the `bulk_download_to_memory` method because we need to keep the order of files, but here we don't need that and I guess using `as_complited` will be faster, but I've not checked yet.

Reply:
Seems like `wait()` would be a more focused choice for this use case. But speaking of downloading performance, I doubt any of these choices would affect it. This is handled by the pool task queue, which executes the submitted tasks after the previous ones are completed (not collected).

https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.wait